### PR TITLE
[WIP] Minimal language experiment

### PR DESCRIPTION
**Pull request by jprochazk**
This PR implements a minimal language which can be used to test out ideas and intial attempts at adding various features.

```rust
let v: int = 10; // int
let v: bool = true; /// bool
let v: str = "test"; /// str
let v: {x:int, y:int} = {x: 10, y: 10}; /// record

let fib n: int -> int =
  if n < 2 then n
  else n * fib (n - 1)
  ;

fib (v.x);

let print_square n: int -> int =
  let r = intToStr (n * n) in
  print (n + " * " + n + " = " + r)
  ;

print_square (v.y);
```

The core language consists of:
- An expression-based syntax
- A very restricted set of types:
  - `int`, `bool`, `str`
  - Structurally-typed records
  - Functions
- Variable expressions
- If-expressions
- Common operators (`+`, `-`, `*`, `/`, `%`, `**`, `==`, `!=`, `<`, `>`, `<=`, `>=`, `&&`, `||`)

The type system will only support unidirectional type inference, and will require the use of type annotations on function parameter and return types.

- [x] Lexer
- [x] Parser
- [ ] Type system
- [ ] Codegen

### CODE REVIEW

File: experiments/type-system/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+struct Args {
+  #[clap(value_parser)]
+  input: PathBuf,
+  #[clap(value_parser)]
+  output: PathBuf,
+}
+
+struct Symbols;
+
+impl ast2str::Symbols for Symbols {
```
Comment (optimalstrategy) [inline experiments/type-system/src/main.rs:23]: There's a built-in thing for this, it's called TestSymbols iirc. Check the parser branch.

**Review by optimalstrategy — COMMENTED** [review]
This is an interesting idea. 

The symbols thing stood out to me, everything else looks okay after a quick scan. Will look at this more comprehensively later.

File: experiments/type-system/src/main.rs (hunk 1 above)
Comment (jprochazk) [reply experiments/type-system/src/main.rs:42]: I tried using `https://insta.rs/` for the snapshot testing, and it's really nice, but it only supported serde as a way to serialize (if I didn't want to use unstructured strings), so I ended up removing the `AstToStr` dependency here

Comment (optimalstrategy) [reply experiments/type-system/src/main.rs:42]: That's fine, but note that insta has [`assert_debug_snapshot`](https://docs.rs/insta/latest/insta/macro.assert_debug_snapshot.html#), and ast2str has [`DisplayAsDebugWrapper`](https://docs.rs/ast2str-lib/1.2.1/ast2str_lib/utils/index.html#example)

**Comment by jprochazk** [comment]
One thing I wanted to clarify about how the parsing works right now: It is not restrictive at all. Any expression can appear anywhere, and that's intentional. I figured you can always come up with pathological cases no matter how hard you try to restrict the grammar, so seeing as everything is an expression, I decided it's easiest to just allow any expression to appear anywhere. For example, you can do this:

```rust
if let check n: int -> bool = n > 10 in check (20)
then print ("n is larger than 20")
else print ("n is not larger than 20")
```

**Pull request closed without merging**